File: src/main/java/howtoprogram/anagrams/AnagramFinder.java
package howtoprogram.anagrams;

import java.io.IOException;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Set;
import java.util.SortedSet;
import java.util.TreeSet;

/**
 * Finds the anagrams of a word.
 */
public class AnagramFinder {

    public static void main(String[] args) throws IOException {
        Path path = Paths.get("src/main/resources/books/english.txt");
        FileBasedDictionary dictionary = new FileBasedDictionary(path);
        String str = "battle";
        if (!dictionary.contains(str)) {
            System.out.println("Not in dictionary: " + str);
        }
        Word word = new Word(str);
        Set<Word> anagrams = word.permutations();
        SortedSet<String> found = new TreeSet<>();
        for (Word w : anagrams) {
            if (dictionary.contains(w.text())) {
                found.add(w.text());
            }
        }
        System.out.println("found = " + found);

    }
}

File: src/main/kotlin/pfb/anagrams/AnagramFinder.kt
package pfb.anagrams

import java.nio.file.Paths


/**
 * Finds the anagrams of a word.
 */
fun main(args: Array<String>) {
    val path = Paths.get("src/main/resources/books/english.txt")
    val dictionary = FileBasedDictionary(path)
    //Create a Word from some initial English word.
    val str = "regal"
    if (!dictionary.contains(str)) {
        System.out.println("Not in dictionary: " + str);
    }
    val word = Term(str);
    //Get all rearrangements of the letters in this word.
    val anagrams = word.permutations()
    //For each of these...
    for (w in anagrams) {
        //...if it is in the dictionary...
        if (dictionary.contains(w.text)) {
            //...print it out.
            println(w.text)
        }
    }
}